### PR TITLE
[Validator] fix URL validator to detect non supported chars according to RFC 3986

### PR DESCRIPTION
**Pull request by e-moe**
| Q             | A
| ------------- | ---
| Branch?       | 3.2
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #21961
| License       | MIT
| Doc PR        | none

### CODE REVIEW

**Comment by Nek-** [comment]
Awesome. Thank you for your contribution. Your code will definitely improve Symfony code base by fixing potential security error and many projects will benefit of that.

I made some tests. This (non-valid) URL isn't validated (which is great) `http://example.com/exploit.html?hel lo`, I think we should add it to the test suite.

I also tested this other URL that is considered as valid but is not: `http://example.com/exploit.html?tçest` this is a problem with many non-ascii chars. Do you plan to fix it? Or your fix (that actually fix some things) will remain as it ?

**Review by Nek- — APPROVED** [review]

**Comment by e-moe** [comment]
@Nek- , thanks for your review! I just extended test cases with your example (`http://example.com/exploit.html?hel lo`). As for non-ascii chars I'm not sure. But previously they were treated as accepted: https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Validator/Tests/Constraints/UrlValidatorTest.php#L104

**Comment by Nek-** [comment]
@e-moe it's valid as DNS but not as query string (where is must be url-encoded). But well, it's fine to me.

**Comment by e-moe** [comment]
oh, I see your point. If needed it can be fixed. Let's wait for other reviews and then we will be able to make a decision

File: src/Symfony/Component/Validator/Constraints/UrlValidator.php [hunk 1]
```diff
@@ -20,8 +20,8 @@
  */
 class UrlValidator extends ConstraintValidator
 {
-    const PATTERN = '~^
-            (%s)://                                 # protocol
+    const PATTERN = '<^
+            (%protocol%)://                         # protocol
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Constraints/UrlValidator.php]: why these two changes? better not do them imho, and continue using sprintf (requires doubling the `%` below)

**Review by nicolas-grekas — COMMENTED** [review]
This should be applied to 2.7, isn't it ? We need to consider if this could introduce any unwanted BC break as side effect.

File: src/Symfony/Component/Validator/Constraints/UrlValidator.php (hunk 1 above)
Comment (e-moe) [inline src/Symfony/Component/Validator/Constraints/UrlValidator.php]: good point, fixed

**Comment by e-moe** [comment]
@nicolas-grekas , I just fixed code based on your comments.

should I create new PR for 2.7 branch?

**Comment by e-moe** [comment]
as for BC break - new constrain is more strict comparing to old version. but new one is closer to RFC

should it be considered as a BC break?

**Comment by Nek-** [comment]
It's not a BC break! It's a fix.

**Comment by fabpot** [comment]
Thank you @e-moe.

**Comment by Aliance** [comment]
I have a *template* link field with Url validator. By template I mean `http://www.domain.com/script.php?value=${PLACEHOLDER}`.
After this pr **Url** validator triggers an error. With the real link it is okay, but that do you recommend for me? The only way is to write my own **UrlTemplate** validator?

**Comment by Nek-** [comment]
@Aliance yes.

I'm sorry to read it breaks your code but this is the attempted behavior of the URL Validator. (here was a potential security issue that is now fixed)

**Pull request closed without merging**